Grammar

nitnelave opened the pull request:
Add grammar, checked by bison, so it is guaranteed not to be ambiguous.
Implement lexer to go with it.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/nitnelave/hopper/19)
<!-- Reviewable:end -->

nitnelave committed:
parser: remove lookahead

nitnelave committed:
grammar: add reference file

nitnelave committed:
examples: update for latest grammar

nitnelave committed:
error: relax constructor constraints

nitnelave committed:
util: introduce LookaheadStack

nitnelave committed:
parser: use LookaheadStack

nitnelave committed:
lexer: use LookaheadStack

nitnelave committed:
lexer: add missing tokens from the grammar

nitnelave committed:
travis: only clone to a depth of 3

nitnelave committed:
coverage: misc

Berdes commented:
A few comments. I didn't reviewed everything thoroughly given the size of the change and that some parts are particularly boring.

---

Reviewed 21 of 21 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[doc/grammar.y, line 1 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosGBqgYnXm4E1GmQI8:-KosGBqgYnXm4E1GmQI9:b-bp0bus) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/doc/grammar.y#L1)):*
> ```Yacc
> %start dummy
> ```

BTW Are we supposed to review this file? It seems really long for this :/

---

*[examples/99_bottles.gh, line 7 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosH6o_B7IdwAvoiaQo:-KosH6o_B7IdwAvoiaQp:b-79yk6r) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/examples/99_bottles.gh#L7)):*
> ```
> /// $n : Number of bottles.
> /// $return : Pretty string.
> fun toBottles(val n: Int) : String {
> ```

Isn't this function pure?

---

*[examples/99_bottles.gh, line 21 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosH33y7uNGy5oo5c1s:-KosH33zMiPANbZXcjxV:bimycmg) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/examples/99_bottles.gh#L21)):*
> ```
> 
> /// Returns the full verse for the number of bottles of beer.
> fun toVerse(val n: Int) =
> ```

Isn't this function also pure?

---

*[examples/Lambdas.gh, line 15 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosGpRW07lPNsUQlg2K:-KosGpRW07lPNsUQlg2L:bz4de8x) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/examples/Lambdas.gh#L15)):*
> ```
>                     // This is equivalent to writing
>                     // map(myLambda)
>                     .map({e -> apply(myLambda, e)})
> ```

I guess it's also equivalent as {e -> myLambda(e)}, right?

---

*[examples/Macros.gh, line 53 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosHKoOPRilwWogbm2y:-KosHKoOPRilwWogbm2z:bce671t) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/examples/Macros.gh#L53)):*
> ```
> ///   this.otherField = otherField;
> /// }
> data_class(class ExampleDataClass {
> ```

Shouldn't this be #data_class?

---

*[examples/Option.gh, line 11 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosHUobsblClUZg6mf8:-KosHUobsblClUZg6mf9:b-c3okgl) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/examples/Option.gh#L11)):*
> ```
> // greater than 3, it returns the value of the input (wrapped in an Option), and
> // otherwise, it returns None (the Option value for nothing).
> fun getInt(val in : Int32) : Int32? = if (in > 3) in else None;
> ```

This function is pure.

---

*[src/error/error.h, line 70 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosIACHgNVUxdreO9DS:-KosIACHgNVUxdreO9DT:b-3dtd50) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/error/error.h#L70)):*
> ```objc
>                 "Error type must be a subclass of Error");
>   template <typename E>
>   using enable_if_error = internals::enable_if_base_of<Err, E>;
> ```

A little bit unrelated to the particular PR, but why are we using template metaprogramming to enforce inheritance when inheritance is not a very recommended pattern in C++?

---

*[src/error/error.h, line 165 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosISVPYXT8QcL0TXni:-KosISVPYXT8QcL0TXnj:bra29r5) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/error/error.h#L165)):*
> ```objc
>   template <typename E>
>   MaybeError(  // NOLINT: explicit
>       E&& value,
> ```

The NOLINT should be on this line.

---

*[src/error/error.h, line 172 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosIVoNTuZSR4m_mfRY:-KosIVoNTuZSR4m_mfRZ:bra29r5) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/error/error.h#L172)):*
> ```objc
>   template <typename E>
>   MaybeError(  // NOLINT: explicit
>       MaybeError<E>&& other,
> ```

The NOLINT should be on this line.

---

*[src/lexer/file_reader.cc, line 6 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosJ2ib2xL8O-wJ6W6d:-KosJ2ib2xL8O-wJ6W6e:b-2yt745) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/lexer/file_reader.cc#L6)):*
> ```cpp
> 
> namespace lexer {
> auto FileReader::read_one_char() -> ErrorOr<State, LexError> {
> ```

Why using trailing return type? It's not necessary nor shorter.

---

*[src/lexer/lexer.cc, line 52 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosK7ktiCtDRwfsxzHX:-KosJvTfbseOUs7gxeOV:b-kqiky8) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/lexer/lexer.cc#L52)):*
> ```cpp
> Lexer::Lexer(const std::string& source, SourceTag tag,
>              const std::string& filename)
>     : reader_{get_stream_from_source(source, tag), filename} {}
> ```

Usually, curly braces are used to construct structs (or simple types) with all the fields as arguments or to construct a container which is supposed to directly contain the arguments. For normal constructors, parenthesis are better.

---

*[src/lexer/lexer.cc, line 88 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosLBCYopqLxstIrHBz:-KosLBCYopqLxstIrHC-:b-ft2t2g) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/lexer/lexer.cc#L88)):*
> ```cpp
> 
>   // Construct a token with the last character.
>   auto make_single_token = [&, this](TokenType tt) {
> ```

Why do you need to capture "this"? Please also capture the variables explicitly. Same for the following lambdas.

---

*[src/lexer/lexer.cc, line 102 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosKzAWqy3KDgx3290i:-KosKzAWqy3KDgx3290j:b-2tzilw) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/lexer/lexer.cc#L102)):*
> ```cpp
>   // corresponding token, if matching. Otherwise, return a token of type
>   // default_type.
>   auto with_second_char = [&, this](
> ```

Given the size of this beast, please consider extracting this as a struct with an operator().

---

*[src/lexer/lexer.cc, line 117 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosM67mBklVYG1kGZnf:-KosM67mBklVYG1kGZng:b-gww4vc) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/lexer/lexer.cc#L117)):*
> ```cpp
>   // LEXER STARTS.
> 
>   switch (current_char()) {
> ```

This whole switch is pretty ugly and looks very error-prone. Do you think there could be a way of simplifying it a little bit using some generalization?

For example, you could have a structure with a list of prefix associated with the action to take (return token and continue, error, call function). Then you would have some piece of code that construct the "state machine" associated and runs it. This code could also verify that you define something for all the cases.

---

*[src/lexer/lexer.h, line 48 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosNfQaYndRWfLYEE0u:-KosNfQaYndRWfLYEE0v:b-tb4sbq) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/lexer/lexer.h#L48)):*
> ```objc
>   // against keywords. Returns either a LOWER_CASE_IDENT or the corresponding
>   // keyword token.
>   ErrorOr<Token, LexError> read_lowercase_identifier();
> ```

Please add #include "lexer/lex_error.h"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nitnelave/hopper/19#-:-KosOQBv1B8TacorhBdg:b-db0p92)*
<!-- Sent from Reviewable.io -->

nitnelave commented:
Addressed the comments in the new PR, since this one is already merged.

---

Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[doc/grammar.y, line 1 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosGBqgYnXm4E1GmQI8:-KosWU9hcMuOIhZnspuJ:bjuzofk) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/doc/grammar.y#L1)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

BTW Are we supposed to review this file? It seems really long for this :/
</blockquote></details>

It's documentation, it's the grammar, you can or not, it's up to you.

---

*[examples/99_bottles.gh, line 7 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosH6o_B7IdwAvoiaQo:-KosWZ2Do0YDPRfLPPuw:bk6ud2y) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/examples/99_bottles.gh#L7)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

Isn't this function pure?
</blockquote></details>

It's not required to mark pure function as such, unless they're needed at compile-time

---

*[examples/99_bottles.gh, line 21 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosH33y7uNGy5oo5c1s:-KosWdUPrh9gG3oOAjdr:bb7od0r) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/examples/99_bottles.gh#L21)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

Isn't this function also pure?
</blockquote></details>

same

---

*[examples/Lambdas.gh, line 15 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosGpRW07lPNsUQlg2K:-KosWfK2VxMzpA13B-0e:b-hcin7i) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/examples/Lambdas.gh#L15)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

I guess it's also equivalent as {e -> myLambda(e)}, right?
</blockquote></details>

Added that.

---

*[examples/Macros.gh, line 53 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosHKoOPRilwWogbm2y:-KosWwBZTgCsfd_9WMLt:b-xd3vm6) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/examples/Macros.gh#L53)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

Shouldn't this be #data_class?
</blockquote></details>

You're right.

---

*[examples/Option.gh, line 11 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosHUobsblClUZg6mf8:-KosXB0Kj-_JhxlX7OBK:b-uv67tg) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/examples/Option.gh#L11)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

This function is pure.
</blockquote></details>

no need to specify

---

*[src/error/error.h, line 70 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosIACHgNVUxdreO9DS:-KosXFPcBmm5m69GdYB7:bvhv0sn) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/error/error.h#L70)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

A little bit unrelated to the particular PR, but why are we using template metaprogramming to enforce inheritance when inheritance is not a very recommended pattern in C++?
</blockquote></details>

There's a difference between "it's overused" and "it's not very recommended". Inheritance in C++ is very recommended when you actually have a is-a relationship, and virtual functions (which is the case for errors here).

---

*[src/error/error.h, line 165 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosISVPYXT8QcL0TXni:-KosXVvET-EgZzTnok3N:bev6nfc) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/error/error.h#L165)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

The NOLINT should be on this line.
</blockquote></details>

Sadly no, it's at the constructor declaration line that clang-tidy warns.

---

*[src/error/error.h, line 172 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosIVoNTuZSR4m_mfRY:-KosXh9NTTcmKRhSrhm7:bb7od0r) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/error/error.h#L172)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

The NOLINT should be on this line.
</blockquote></details>

same

---

*[src/lexer/file_reader.cc, line 6 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosJ2ib2xL8O-wJ6W6d:-KosXpLvAfHkNbVSATje:bm1my4j) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/lexer/file_reader.cc#L6)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

Why using trailing return type? It's not necessary nor shorter.
</blockquote></details>

`State` is resolve as `FileReader::State` in the trailing return type style, not in the usual one.
I toyed with the idea of making it the style of the repository to have trailing return style to match grasshopper, but ended up against as it was too much of a departure from usual C++ style

---

*[src/lexer/lexer.cc, line 52 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosK7ktiCtDRwfsxzHX:-KosYDvHQWQIcIGVaxhm:bdapclv) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/lexer/lexer.cc#L52)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

Usually, curly braces are used to construct structs (or simple types) with all the fields as arguments or to construct a container which is supposed to directly contain the arguments. For normal constructors, parenthesis are better.
</blockquote></details>

Although I could argue for braces almost everywhere (like I could argue for almost always auto), I'll give in for consistency. But I disagree on principle with the "parenthesis are better" statement

---

*[src/lexer/lexer.cc, line 88 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosLBCYopqLxstIrHBz:-KosYWlcRfJha91ee5SZ:b-wtdvnu) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/lexer/lexer.cc#L88)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

Why do you need to capture "this"? Please also capture the variables explicitly. Same for the following lambdas.
</blockquote></details>

current should be `this->current_token()`, let me replace that.
I don't see the point in capturing variables explicitly for lambdas, especially ones used only locally as here.

---

*[src/lexer/lexer.cc, line 102 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosKzAWqy3KDgx3290i:-KosZYr2bZTYYNi8WlGx:b-896fix) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/lexer/lexer.cc#L102)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

Given the size of this beast, please consider extracting this as a struct with an operator().
</blockquote></details>

Done.

---

*[src/lexer/lexer.cc, line 117 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosM67mBklVYG1kGZnf:-KosbVDe3lruU0GlLioT:bshrbj) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/lexer/lexer.cc#L117)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

This whole switch is pretty ugly and looks very error-prone. Do you think there could be a way of simplifying it a little bit using some generalization?

For example, you could have a structure with a list of prefix associated with the action to take (return token and continue, error, call function). Then you would have some piece of code that construct the "state machine" associated and runs it. This code could also verify that you define something for all the cases.
</blockquote></details>

I'm open to refactors :)
But for now, this is a well-tested piece of code, easily extensible, and I find it easily readable. Once you find the char you're interested in, you never have more than 10 lines of linear flow, compared to a state machine that would have you jump around.

Maybe you have a more elegant idea in mind.

Ideally, I would (statically or at startup) construct a trie with all fixed keywords, and just have rules for the rest, but for now it works, it's well contained, well tested, maintainable, let's move forward :)

---

*[src/lexer/lexer.h, line 48 at r1](https://reviewable.io:443/reviews/nitnelave/hopper/19#-KosNfQaYndRWfLYEE0u:-Koscye9y9tgxPZkGuyz:b-896fix) ([raw file](https://github.com/nitnelave/hopper/blob/b985d44ed5338e97d510975dc5e1f6e9a2cbe9f9/src/lexer/lexer.h#L48)):*
<details><summary><i>Previously, Berdes wrote…</i></summary><blockquote>

Please add #include "lexer/lex_error.h"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nitnelave/hopper/19#-:-Kosd-Jt0AmsRYnKA6vg:b-4o5vh7)*
<!-- Sent from Reviewable.io -->
